paparazzi/src/main/java/app/cash/paparazzi/internal/compose/RecomposerWatcher.kt

```diff
@@ -0,0 +1,86 @@
+@file:OptIn(ExperimentalComposeUiApi::class)
+
+package app.cash.paparazzi.internal.compose
+
+import androidx.compose.runtime.Recomposer
+import androidx.compose.ui.ExperimentalComposeUiApi
+import androidx.compose.ui.InternalComposeUiApi
+import androidx.compose.ui.platform.WindowRecomposerPolicy
+import androidx.compose.ui.platform.createLifecycleAwareWindowRecomposer
+import app.cash.paparazzi.PaparazziSdk.Companion.isPresentInClasspath
+import app.cash.paparazzi.internal.PaparazziLogger
+import com.android.ide.common.rendering.api.Result
+import com.android.ide.common.rendering.api.Result.Status.ERROR_UNKNOWN
+import kotlin.coroutines.CoroutineContext
+
+@InternalComposeUiApi
+internal class RecomposerWatcher(
+  private val mainDispatcher: CoroutineContext,
+  private val paparazziLogger: PaparazziLogger
+) {
+
+  private var recomposer: Any? = null
+  private var frameLimit: Int = 1 // TODO: Configure with Test rule
+  private var additionalRenderLimit: Int = 1 // TODO: Configure with TestRule
+
+  fun register() {
+    // By default, Compose UI uses its own implementation of CoroutineDispatcher, `AndroidUiDispatcher`.
+    // Since this dispatcher does not provide its own implementation of Delay, it will default to using DefaultDelay which runs
+    // async to our test Handler. By initializing Recomposer with Dispatchers.Main, Delay will now be backed by our test Handler,
+    // synchronizing expected behavior.
+    if (hasComposeRuntime) {
+      WindowRecomposerPolicy.setFactory {
+        val windowRecomposer = it.createLifecycleAwareWindowRecomposer(mainDispatcher)
+        recomposer = windowRecomposer
+        return@setFactory windowRecomposer
+      }
+    }
+  }
+
+  private fun renderForIdleRecomposer(
+    frame: Int,
+    additionalRenderCount: Int,
+    result: Result,
+    onAdditionalRender: () -> Result
+  ): Result {
+    if (hasComposeRuntime && result.status != Result.Status.ERROR_UNKNOWN) {
+      // If there is a recomposition that needs to happen, we need to trigger it within the context of the first frame.
+      val recomposerInstance = recomposer as? Recomposer
+      val hasPendingWork = recomposerInstance?.hasPendingWork
+      if (frame < frameLimit && hasPendingWork == true) {
+        if (additionalRenderCount >= additionalRenderLimit) {
+          paparazziLogger.info(
```

Comment:
What are your thoughts on making this an error? If we're taking a single snapshot and the view never stabilizes an error would be a nice debugging tool. With the info log it could be ignored / missed in all the output.

Maybe we could provide a flag that makes it an error or not?

Reply:
hmm, maybe I can up it to a warning. I wasn't able to hit this via a repro case but its more of insight if this does occur.